File: app/src/main/java/eu/letmehelpu/android/ConversationActivity.java
package eu.letmehelpu.android;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class ConversationActivity extends AppCompatActivity {

    MessagesAdapter adapter = new MessagesAdapter();
    long userId = 9;
    private Conversation document;
    private String TAG = this.getClass().getSimpleName();
    private ListenerRegistration registration;

    public static Intent createIntent(Context context, Conversation conversation) {
        Intent intent = new Intent(context, ConversationActivity.class);
        intent.putExtra("conversationDocumentId", conversation);
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        document = (Conversation) getIntent().getSerializableExtra("conversationDocumentId");

        setContentView(R.layout.conversation);

        FirebaseFirestore db = FirebaseFirestore.getInstance();
        registration = db.collection("conversations")
                .document(document.documentId)
                .collection("messages")
                //.orderBy("timestamp")
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
            @Override
            public void onEvent(@javax.annotation.Nullable QuerySnapshot queryDocumentSnapshots, @javax.annotation.Nullable FirebaseFirestoreException e) {
                List<Message> messages = queryDocumentSnapshots.toObjects(Message.class);
                Collections.sort(messages, new Comparator<Message>() {
                    @Override
                    public int compare(Message o1, Message o2) {
                        return -Float.compare(o1.timestamp, o2.timestamp);
                    }
                });
                adapter.setMessages(messages);
            }
        });

        RecyclerView messages = findViewById(R.id.messages);
        messages.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, true));
        messages.setAdapter(adapter);

        EditText messageInput = findViewById(R.id.message_input);
        messageInput.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_SEND) {
                    sendMessage(v.getText().toString());
                    v.setText("");
                    return true;
                }
                return false;
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        registration.remove();
    }

    public void sendMessage(final String text) {
        final Message message = new Message();
        message.seen = false;
        message.text = text;
        message.timestamp = System.currentTimeMillis();
        final FirebaseFirestore db = FirebaseFirestore.getInstance();
        db.collection("conversations").document(document.documentId)
                .collection("messages")
                .add(message)
                .addOnSuccessListener(this, new OnSuccessListener<DocumentReference>() {
                    @Override
                    public void onSuccess(DocumentReference documentReference) {
                        document.lastMessage = text;
                        document.lastMessageBy = userId;
                        document.timestamp =  message.timestamp;

                        db.collection("conversations")
                                .document(document.documentId)
                                .set(document)
                                .addOnCompleteListener(ConversationActivity.this, new OnCompleteListener<Void>() {
                                    @Override
                                    public void onComplete(@NonNull Task<Void> task) {
                                        Log.d(TAG, "onComplete() called with: task = [" + task + "]");
                                        Toast.makeText(ConversationActivity.this, "SENDED", Toast.LENGTH_SHORT).show();
                                    }
                                });
                    }
                });


    }
}

File: app/src/main/java/eu/letmehelpu/android/conversation/ConversationActivity.kt
package eu.letmehelpu.android.conversation

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.TextView
import dagger.android.support.DaggerAppCompatActivity
import eu.letmehelpu.android.R
import eu.letmehelpu.android.messaging.MessagingService
import eu.letmehelpu.android.messaging.SendMessage
import eu.letmehelpu.android.model.Conversation
import javax.inject.Inject

class ConversationActivity : DaggerAppCompatActivity() {

    private lateinit var adapter : MessagesListAdapter
   // private lateinit var userId: Long
    private lateinit var conversation: Conversation
    private lateinit var conversationViewModel: ConversationViewModel

    @Inject
    lateinit var sendMessage: SendMessage

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val userId = intent.getLongExtra(EXTRA_USER_ID, -1)
        conversation = intent.getSerializableExtra(EXTRA_CONVERSATION) as Conversation
        val otherUsers = conversation.users.keys.map { it -> it.toLong() }.filter { it != userId }
        adapter = MessagesListAdapter(otherUsers)

        setContentView(R.layout.conversation)

        val messages = findViewById<RecyclerView>(R.id.messages)
        messages.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, true)
        messages.adapter = adapter


        conversationViewModel = ViewModelProviders.of(this, ConversationViewModelFactory(userId, conversation, sendMessage)).get(ConversationViewModel::class.java)

        conversationViewModel.getMessages().observe(this, Observer {
            it?.let{ adapter.submitList(it)}
        })

        conversationViewModel.getUserIdsToReadTimes().observe(this, Observer {
            it?.let { adapter.setLastReaded(it) }
        })

        val messageInput = findViewById<EditText>(R.id.message_input)
        messageInput.setOnEditorActionListener(TextView.OnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                conversationViewModel.sendMessage(v.text.toString())
                v.text = ""
                return@OnEditorActionListener true
            }
            false
        })
    }

    override fun onStart() {
        super.onStart()
        startedConversation = conversation.documentId
        sendBroadcast(MessagingService.createDeleteNotificationIntent(this, conversation))
    }

    override fun onStop() {
        super.onStop()
        startedConversation = null
    }

    companion object {
        var startedConversation:String? = null
        private val EXTRA_CONVERSATION = "EXTRA_CONVERSATION"
        private val EXTRA_USER_ID = "EXTRA_USER_ID"

        fun createIntent(context: Context, userId: Long, conversation: Conversation): Intent {
            val intent = Intent(context, ConversationActivity::class.java)
            intent.putExtra(EXTRA_CONVERSATION, conversation)
            intent.putExtra(EXTRA_USER_ID, userId)
            return intent
        }
    }
}
